pkg/controller/mizarcontrollers/mizar-pod-controller.go

```diff
@@ -0,0 +1,194 @@
+
+	return oc
+}
+
+// Run begins watching and handling.
+func (oc *PodController) Run(workers int, stopCh <-chan struct{}) {
+	defer utilruntime.HandleCrash()
+	defer oc.queue.ShutDown()
+
+	klog.Infof("Starting %v controller", controllerFor)
+	defer klog.Infof("Shutting down %v controller", controllerFor)
+
+	if !controller.WaitForCacheSync(controllerKind, stopCh, oc.listerSynced) {
+		return
+	}
+
+	for i := 0; i < workers; i++ {
+		go wait.Until(oc.worker, time.Second, stopCh)
+	}
+
+	<-stopCh
+}
+
+func (oc *PodController) createObj(obj interface{}) {
+	key, _ := controller.KeyFunc(obj)
+	oc.queue.Add(KeyWithEventType{Key: key, EventType: "create"})
+}
+
+// When an object is updated.
+func (oc *PodController) updateObj(old, cur interface{}) {
+	curObj := cur.(*v1.Pod)
+	oldObj := old.(*v1.Pod)
+	if curObj.ResourceVersion == oldObj.ResourceVersion {
+		// Periodic resync will send update events for all known objects.
+		// Two different versions of the same object will always have different RVs.
+		return
+	}
+	if curObj.DeletionTimestamp != nil {
+		// Object is being deleted. Don't handle update anymore.
+		return
+	}
+
+	key, _ := controller.KeyFunc(curObj)
+	oc.queue.Add(KeyWithEventType{Key: key, EventType: "update"})
+}
+
+func (oc *PodController) deleteObj(obj interface{}) {
+	key, _ := controller.KeyFunc(obj)
+	klog.Infof("%v deleted. key %s.", controllerFor, key)
+	oc.queue.Add(KeyWithEventType{Key: key, EventType: "delete"})
+}
+
+// worker runs a worker thread that just dequeues items, processes them, and marks them done.
+// It enforces that the handler is never invoked concurrently with the same key.
+func (oc *PodController) worker() {
+	for oc.processNextWorkItem() {
+	}
+}
+
+func (oc *PodController) processNextWorkItem() bool {
+	workItem, quit := oc.queue.Get()
+
+	if quit {
+		return false
+	}
+
+	keyWithEventType := workItem.(KeyWithEventType)
+	key := keyWithEventType.Key
+	eventType := keyWithEventType.EventType
+	defer oc.queue.Done(key)
+
+	err := oc.handler(key, eventType)
+	if err == nil {
+		oc.queue.Forget(key)
+		return true
+	}
+
+	utilruntime.HandleError(fmt.Errorf("Handle %v of key %v failed with %v", controllerFor, key, err))
+	oc.queue.AddRateLimited(keyWithEventType)
+
+	return true
+}
+
+func (oc *PodController) handle(key string, eventType string) error {
+	klog.Infof("Entering handling for %v. key %s", controllerFor, key)
+
+	startTime := time.Now()
+	defer func() {
+		klog.V(4).Infof("Finished handling %v %q (%v)", controllerFor, key, time.Since(startTime))
+	}()
+
+	tenant, namespace, name, err := cache.SplitMetaTenantNamespaceKey(key)
+	if err != nil {
+		return err
+	}
+
+	obj, err := oc.lister.PodsWithMultiTenancy(namespace, tenant).Get(name)
+	if errors.IsNotFound(err) {
+		klog.V(4).Infof("%v %v cannot be found", controllerFor, key)
+		return nil
+	} else {
+		klog.V(4).Infof("Handling %v %s/%s/%s hashkey %v for event %v", controllerFor, tenant, namespace, obj.Name, obj.HashKey, eventType)
```

Comment:
Isuggest to add logic of processing potential errors

Reply:
+1

Errors other than NotFound are not handled.

Reply:
Nice catch. Actually IsNotFound is not special here. Original I reserved it for deletion scenario but actually it doesn't apply. I changed code to return error when it happens. After error returns, it will be logged and retried.